source/kernel/src/registry/mod.rs

```diff
@@ -0,0 +1,440 @@
+        Some(UserspaceHandle {
+            req_producer_leaked: item.value.req_prod_leaked.clone(),
+            req_deser: item.value.req_deser?,
+        })
+    }
+}
+
+// UserRequest
+
+// UserResponse
+
+impl<U: MaxSize> MaxSize for UserResponse<U> {
+    const POSTCARD_MAX_SIZE: usize = {
+        <[u8; 16] as MaxSize>::POSTCARD_MAX_SIZE
+            + <u32 as MaxSize>::POSTCARD_MAX_SIZE
+            + <Result<U, ()> as MaxSize>::POSTCARD_MAX_SIZE
+    };
+}
+
+// Envelope
+
+// Message
+
+// ReplyTo
+
+impl<U> ReplyTo<U> {
+    pub async fn reply_konly(self, payload: Result<U, ()>) -> Result<(), ()> {
+        let hmsg = Envelope { body: payload };
+        match self {
+            ReplyTo::KChannel(kprod) => kprod.enqueue_async(hmsg).await.map_err(drop),
+            ReplyTo::Rosc(sender) => sender.send(hmsg),
+            ReplyTo::Userspace { .. } => Err(()),
+        }
+    }
+}
+
+impl<U> ReplyTo<U>
+where
+    U: Serialize + MaxSize,
+{
+    pub async fn reply(self, uuid_source: Uuid, payload: Result<U, ()>) -> Result<(), ()> {
+        match self {
+            ReplyTo::KChannel(kprod) => {
+                let hmsg = Envelope { body: payload };
+                kprod.enqueue_async(hmsg).await.map_err(drop)
+            }
+            ReplyTo::Rosc(sender) => {
+                let hmsg = Envelope { body: payload };
+                sender.send(hmsg)
+            }
+            ReplyTo::Userspace { nonce, outgoing } => {
+                let mut wgr = outgoing
+                    .send_grant_exact(<UserResponse<U> as MaxSize>::POSTCARD_MAX_SIZE)
+                    .await;
+                let used = postcard::to_slice(
+                    &UserResponse {
+                        uuid: uuid_source,
+                        nonce,
+                        reply: payload,
+                    },
+                    &mut wgr,
+                )
+                .map_err(drop)?;
+                let len = used.len();
+                wgr.commit(len);
+                Ok(())
+            }
+        }
+    }
+}
+
+// UserspaceHandle
+
+impl UserspaceHandle {
+    pub fn process_msg(
+        &self,
+        user_msg: UserRequest<'_>,
+        user_ring: &bbq::MpscProducer,
+    ) -> Result<(), ()> {
+        unsafe { (self.req_deser)(user_msg, &self.req_producer_leaked, user_ring) }
+    }
+}
+
+// KernelHandle
+
+impl<RD: RegisteredDriver> KernelHandle<RD> {
+    pub async fn send(&self, msg: RD::Request, reply: ReplyTo<RD::Response>) -> Result<(), ()> {
+        self.prod
+            .enqueue_async(Message {
+                msg: Envelope { body: msg },
+                reply,
+            })
+            .await
+            .map_err(drop)
+    }
+}
+
+// -- other --
+
+/// A monomorphizable function that allows us to store the serialization type within
+/// the function itself, allowing for a type-erased function pointer to be stored
+/// inside of the registry.
```

Comment:
nit: would be nice if there was a mention of this fn's safety invariants in this comment?